pkg/bgpv1/manager/store/diffstore.go

```diff
@@ -115,22 +121,30 @@ func (ds *diffStore[T]) handleEvent(event resource.Event[T]) {
 	event.Done(nil)
 }
 
-// Diff returns a list of items that have been upserted(updated or inserted) and deleted since the last call to Diff.
-func (ds *diffStore[T]) Diff() (upserted []T, deleted []resource.Key, err error) {
+// Diff returns a list of items that have been upserted (updated or inserted) and deleted since the last call
+// to Diff with the given callerID.
+func (ds *diffStore[T]) Diff(callerID string) (upserted []T, deleted []resource.Key, err error) {
 	ds.mu.Lock()
 	defer ds.mu.Unlock()
 
 	if ds.store == nil {
 		return nil, nil, ErrStoreUninitialized
 	}
 
+	updatedKeys, ok := ds.callerUpdatedKeys[callerID]
+	if !ok {
+		// caller-specific updated keys map does not exist yet, create and return all items as upserted
+		ds.callerUpdatedKeys[callerID] = make(updatedKeysMap)
```

Comment:
While we create the `callerUpdatedKeys` entry here, we don't have any logic to remove the entry. This results in leaking the memory for the diffs for the BGP instance that no longer exists. Also, when we delete and re-add the BGP instance, it potentially gets some garbage diffs.

In the current design, ServiceReconciler keeps the per-instance state locally. I think this causes this leakage. This is the design we declined in the past and we decided to keep the per-instance state in the metadata. Maybe it's time for us to revisit this?

Reply:
Ah right, looks like we may need to introduce something like a new `Cleanup(instance)` method for the reconcilers to cleanup per-instance state held outside of the reconciler metadata.

Reply:
+1 to introduce `Cleanup(instance)`.